packages/next/next-server/server/image-optimizer.ts

```diff
@@ -0,0 +1,162 @@
+import { createHash } from 'crypto'
+import Server from './next-server'
+import { fileExists } from '../../lib/file-exists'
+
+let sharp: typeof import('sharp')
+const AVIF = 'image/avif'
+const WEBP = 'image/webp'
+const MEDIA_TYPES = [/* AVIF, */ WEBP]
+
+export async function imageOptimizer(
+  server: Server,
+  req: IncomingMessage,
+  res: ServerResponse
+) {
+  const { nextConfig, distDir } = server
+  const { sizes = [], domains = [] } = nextConfig?.experimental?.images || {}
+  const { url: reqUrl = '/', headers } = req
+  const { query } = parse(reqUrl, true)
+  const { url, w, q } = query
+  const proto = headers['x-forwarded-proto'] || 'http'
+  const host = headers['x-forwarded-host'] || headers.host
+  const mediaType = accept.mediaType(req.headers.accept, MEDIA_TYPES)
+
+  if (!url) {
+    res.statusCode = 400
+    res.end('"url" parameter is required')
+    return { finished: true }
+  } else if (Array.isArray(url)) {
+    res.statusCode = 400
+    res.end('"url" parameter cannot be an array')
+    return { finished: true }
+  }
+
+  let absoluteUrl: URL
+  try {
+    absoluteUrl = new URL(url)
+
+    if (
+      Array.isArray(domains) &&
+      domains.length > 0 &&
+      !domains.includes(absoluteUrl.hostname)
+    ) {
+      res.statusCode = 400
+      res.end('"url" parameter is not allowed')
+      return { finished: true }
+    }
+  } catch (_error) {
+    // url was not absolute so assuming relative url
+    try {
+      absoluteUrl = new URL(url, `${proto}://${host}`)
+    } catch (__error) {
+      res.statusCode = 400
+      res.end('"url" parameter is invalid')
+      return { finished: true }
+    }
+  }
+
+  if (!w) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) is required')
+    return { finished: true }
+  } else if (Array.isArray(w)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) cannot be an array')
+    return { finished: true }
+  }
+
+  if (!q) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) is required')
+    return { finished: true }
+  } else if (Array.isArray(q)) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) cannot be an array')
+    return { finished: true }
+  }
+
+  const width = parseInt(w)
+
+  if (!width || isNaN(width)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) must be a number greater than 0')
+    return { finished: true }
+  }
+
+  if (Array.isArray(sizes) && sizes.length > 0 && !sizes.includes(width)) {
+    res.statusCode = 400
+    res.end(`"w" parameter (width) of ${width} is not allowed`)
+    return { finished: true }
+  }
+
+  const quality = parseInt(q)
+
+  if (isNaN(quality) || quality < 1 || quality > 100) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) must be a number between 1 and 100')
+    return { finished: true }
+  }
+
+  const { href } = absoluteUrl
+  const fileName = getFileName([href, width, quality, mediaType])
```

Comment:
We should potentially also consider a set value like `a` that can be incremented when a change to the optimization has been made. This would be similar to what we do for Babel and Terser caches.

Reply:
I added CACHE_VERSION=1 in 3257808fe4432f9a1e2cd82bd99191f73445f291

Is that what you're looking for?